### ISSUE TITLE
Improve suggested mod names when installing downloads

### ISSUE THREAD

**Issue by AnyOldName3** [ModOrganizer2/modorganizer#742]
# Problem

The current behaviour is to just use the Nexus mod name (or maybe the name the mod had when the first file from it was installed or something like that). This is fine for the *I'll just press merge in any situation and then reinstall optional files if I have to* approach, but users (including @isanae) don't use this approach and were suggesting changing it to add the download name to the suggested name, too (potentially with separate handling for main files, updates and optional files). I don't like that solution, as it breaks existing workflows and basically forces everyone into whatever workflow whichever developer implements it likes.

# Solution

Maybe we can have a setting that controls the proposed name when installing a new file with machine-replaceable parameters like `%nexusmodname%` and `%downloadname%`. That way, I could set it to `$if (%previousversioninstalled%, %previousversionmod%) $else (%nexusmodname%)` and someone else could set it to `%nexusmodname% - %downloadname% - %version%`, and a third person could set it to `$if ( $or (%ismainfile%, %isupdate%), %nexusmodname%) $else (%nexusmodname% - %downloadname%)`.
MusicBrainz Picard has the conditional stuff sorted already for its file naming feature and is GPL 2+, so we can steal their implementation, and it's going to have nicer syntax than whatever we can come up with without copying.

We should default it to something that replicates the current behaviour as that means people who don't care can just keep on clicking without thinking. We might also want to provide a few examples of popular options so that people don't have to do everything themselves and can pick something resembling what they like and tweak it if necessary.

# Describe alternatives you've considered
* Leaving things the same - annoys lots of people, doesn't break anyone's existing workflow or muscle memory.
* Implementing a new fixed behaviour - helps people who like that one specific fixed behaviour and disrupts the workflow of everyone else.
* Implementing lots of fixed behaviours and using them to populate the drop-down list we already have - probably more work than my solution, doesn't give quite as much freedom, and still requires people to click buttons they don't want to.
* Letting a plugin handle it - probably actually harder to implement and scarier for users to customise.

# Additional context

I've already decided my idea is the best.

**Comment by AnyOldName3** [comment]
Here's a link to a mostly-non-scary part of MusicBrainz Picard's documentation on its system: https://picard.musicbrainz.org/docs/scripting/#file_naming_examples

**Comment by AnyOldName3** [comment]
@Al12rs proposed a new UI like this:

https://cdn.discordapp.com/attachments/412414009694748693/584690345221292042/unknown.png

Advantages of AL's solution to the mod install thing - you're only one click away from doing something that makes some kind of sense to someone.

Disadvantages:

* Which name do we propose? If it's the first file from a mod, do we just use the mod name, or do we always append the download name to be consistent? Are we going to have people who merge complaining that they have the mod *CBBE - Main File - v2.3.4* that actually contains version 4.5.6 and some optional files or people who install to separate mods complaining that they have one mod called *CBBE* and the rest are all `CBBE - <download> - <info>` and that's inconsistent?
* The UI has lots of options and is going to be scary and confusing to new users. At least with the old UI, we could just tell new users that they could click merge and it would basically behave like manual modding or NMM like they're used to.
* If the proposed mod name isn't right, there's a need for clicking or typing. I'd expect that any given user is going to stick to one system for 99% of their mods and that any such system can be described with my proposed solution and then mean that 99% of mod installs are just one click.

I'm still thinking my solution (as described in the issue report I opened before reading this channel) is better.

**Comment by Al12rs** [comment]
I disagree on the fact that users will behave consistently, always making new mods or always merging.
A very simple example: I always create a new mod, never merge. But I see that there is a new mod update after a week I didn't mod, I download that mod and install it. When installing I Mo2 asks if I want to overwrite and I realize I had the previous version I had forgotten about, I press overwrite. 

People might rely on the current system to check whether they have already a mod installed from that page. I don't actually remember my mods in a 150 mods setup, I doubt people with 1k setups do either.

My proposition allows users to get the information of what mods they already have installed and a quick way to either merge or overwrite. 

What is missing is a good  automatic way to generate alternative names, for which the system proposed by @AnyOldName3 could be used, or some other variation. 

I think the new dialog can be implemented even without the a good name system as the situation for people who always create new mods is not worse as before they had to change the name (I usually found a good enough name from the drop down and maybe edited it a little. 

About the dialog being too hard for filthy casuals, I believe it's better to make them understand what they are doing. 

Tr;Dr: I think the new install dialog and the new name system could happily coexist, but I believe the namesystem alone is not enough.

**Comment by AnyOldName3** [comment]
The system you've proposed doesn't create a *95% of the time, you just click one button* situation, and that's really the whole point in my proposal. If we did *both*, we could use that to automatically generate the suggested name for a new mod and to select the right existing mod in the list of already installed mods, and get the *it's usually just one click and no typing* thing.

> I disagree on the fact that users will behave consistently, always making new mods or always merging.

They don't need to be 100% consistent, just usually consistent. I wasn't suggesting that we ditch the *pick a mod/name* and *merge, replace, rename* dialogue in favour of always doing one thing without confirmation, just that we fill the mod/name box automatically with something the user likes, so most of the time, they can just press a single button when that dialogue appears.

> People might rely on the current system to check whether they have already a mod installed from that page.

I hadn't considered this because I'd misremembered how the dialogue worked and so forgot that it wouldn't tell you this if you picked a different name. There should probably be a better way of doing that, as it kind of feels like checking if your work is saved by clicking the close button and seeing if you get a warning (in that you're putting yourself a single click away from messing something up that normally requires several). Either way, this only gets broken by my proposal because the current dialogue is shit, not because my proposal is bad.

> and a quick way to either merge or overwrite.

The issue is that you're going to need to look at what mods you've got and pick a mod to merge into or overwrite in addition to just pressing merge or overwrite, instead of having the option you want pre-selected automatically so it's just one click to confirm.

> I think the new dialog can be implemented even without the a good name system as the situation for people who always create new mods is not worse as before they had to change the name (I usually found a good enough name from the drop down and maybe edited it a little.

It doesn't, however, make things any better for anyone else, but makes it look like the problem is solved, so will probably make any further improvements a much lower priority, and then they'll get put off forever.

tl;dr: I think the new install dialogue and the new name system could happily coexist, but I believe the dialogue alone is not enough.

**Comment by Al12rs** [comment]
That was well argued, I agree, we should try to make both proposal happen or the desired goal will not be reached.

Some remarks though:

> The issue is that you're going to need to look at what mods you've got and pick a mod to merge into or overwrite in addition to just pressing merge or overwrite, instead of having the option you want pre-selected automatically so it's just one click to confirm.

We could make the list of related mod already select the one that would have been overwritten by the current system so the brain dead users can press the merge button.
What I'm saying though is that do we actually want to have people mindlessly merging their mods? No we what people to be able to competently decide what they what to do based on useful info. Where currently users had no way of knowing whether there even was already a mod installed unless they install another one with the same name and that dialog shows up, we want people to be able to see all the related mods so they can decide whether it's a merge with its an overwriting update to the main mod, a patch they want to keep separate, an update to said patch etc. I really want the user to be able to see all the related mods and decide what to do.

**Comment by AnyOldName3** [comment]
The main annoyance I *personally* have is that for the few mods I have where I *don't* just merge optional files, most of the time I've clicked through the dialogue completely on autopilot and replaced the main file with the optional file or vice versa. Most of the time, though, I don't want optional files kept separately as I'm more likely to run into issues by enabling or disabling half of a mod if there's no shared checkbox than I am to mess up the installation order of the optional files (and I'm genuinely too lazy to click on a dropdown and click again to select an alternate name when updating lots of mods - do the new dialogue really well and that might go away, though).

In my particular case, if a mod is an update for another file, defaulting the selection to the mod that other file is in would resolve my main complaint. However, I know there are people who install new versions of existing files to a new mod, and implementing this approach wouldn't help them out at all.

**Comment by Al12rs** [comment]
Well if I'm getting it right neither of us is going to work on either feature correct? Well at least I can't until december or so. So we hope in @isa?

**Comment by AnyOldName3** [comment]
I won't be doing it immediately, and that means I have no way of telling if and when I might get around to it. Sometimes I say I won't do something and that someone else should, and then if it takes more than an hour for someone else to start, I end up doing it myself. That's how the FNIS integration plugin came about, for example.

**Comment by isanae** [comment]
This is probably the main thing I've always wanted to improve, so yes I intend to work on it. I'll go through everything that was said here and on discord and think about the next step. It's more complex than what I've worked on until now and has consequences on workflows that are much more important than my puny UI changes.

**Comment by Al12rs** [comment]
Instead of this dialog https://cdn.discordapp.com/attachments/412414009694748693/584690345221292042/unknown.png , to show the user which mods are already installed we could instead use the existing dropdown, just making it obvious which are suggested new names and which are existing mods, so that the user can do a conscious choice.
